README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# PLreg

<!-- badges: start -->
<!-- badges: end -->

The **PLreg** package allows fitting power logit regression models.
Diagnostic tools associated with the fitted model, such as the
residuals, local influence measures, leverage measures, and goodness-of-
fit statistics, are implemented.

## Installation

You can install the development version of PLreg from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ffqueiroz/PLreg")
```

## Main functions

#### <tt>dPL</tt>, <tt>pPL</tt>, <tt>qPL</tt>, and <tt>rPL</tt>

Currently, the
![\\textbf{PLreg}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Ctextbf%7BPLreg%7D "\textbf{PLreg}")
package includes 7 members of the power logit class of distributions:
the power logit normal, power logit Student-t, power logit type II
logistic, power logit power exponential, power logit sinh-normal, power
logit hyperbolic and power logit slash distributions. The package
provides the <tt>dPL</tt>, <tt>pPL</tt>, and <tt>qPL</tt> functions to
compute the probability density function, cumulative distribution
function and quantile function of the power logit distribution. Also,
the <tt>rPL</tt> function may be used to generate random samples from
power logit distributions. The basic usages of these functions are:

``` r
dPL(x, mu, sigma, lambda, zeta = 2, family, log = FALSE)

pPL(q, mu, sigma, lambda, zeta = 2, family, lower.tail = TRUE, log.p = FALSE)

qPL(p, mu, sigma, lambda, zeta = 2, family, lower.tail = TRUE, log.p = FALSE)

rPL(n, mu, sigma, lambda, zeta = 2, family)
```

#### <tt>PLreg</tt>

The main function of the
![\\textbf{PLreg}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Ctextbf%7BPLreg%7D "\textbf{PLreg}")
package is <tt>PLreg()</tt>, which allows to fitting power logit
regression model to proportional data; this explains the name. The
arguments of <tt>PLreg()</tt> are:

``` r
PLreg(formula, data, subset, na.action, family = c("NO", "LO", "TF", "PE", "SN", "SLASH", "Hyp"), 
      zeta = NULL, link = c("logit", "probit", "cloglog", "cauchit", "log", "loglog"), 
      link.sigma = NULL, type = c("pML", "ML"), control = PLreg.control(...), 
      model = TRUE, y = TRUE, x = FALSE, ...)
```

The <tt>PLreg()</tt> function returns an object of class
“<tt>PLreg</tt>”, similar to “<tt>betareg</tt>” and “<tt>glm</tt>”
objects, for which some methods available. The <tt>summary()</tt> method
returns a standard output, with coefficient estimates, standard errors,
partial Wald tests and p values for the regression coefficients, the
overall goodness-of-fit measure, the pseudo
![R^2](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;R%5E2 "R^2"),
etc.. The <tt>type</tt> argument in <tt>summary()</tt> specifies the
type of residuals included in the output; currently three residuals are
supported: <tt>“standardized”</tt>, </tt>“quantile”</tt> and
<tt>“deviance”</tt>. The <tt>plot()</tt> method draws graphs for
diagnostic and influence analyses.

#### <tt>extra.parameter</tt>

An important function in the
![\\textbf{PLreg}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Ctextbf%7BPLreg%7D "\textbf{PLreg}")
package is <tt>extra.parameter()</tt>. It can be used to estimate the
extra parameter of some power logit models. The basic usage is:

``` r
extra.parameter(object, lower, upper, grid = 10)
```

## Example

``` r
library(PLreg)
#> 
#> Attaching package: 'PLreg'
#> The following object is masked from 'package:stats':
#> 
#>     influence
## basic example code
```

In the following, an example is presented to illustrate the capacities
of
![\\textbf{PLreg}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Ctextbf%7BPLreg%7D "\textbf{PLreg}")
package. We use the <tt>bodyfat_Aeolus</tt> data set, available in the
package.

    help(bodyfat_Aeolus, package = "PLreg")

The response variable is <tt>percentfat</tt> and the covariates are the
sex of the sampled bat (<tt>sex</tt>), the hibernation time
(<tt>days</tt>) and the year that the bat was sampled (<tt>year</tt>).
We start by fitting a power logit power exponential regression model
with constant dispersion and
![\\zeta = 2](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Czeta%20%3D%202 "\zeta = 2").
To select a suitable value for
![\\zeta](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Czeta "\zeta"),
use the <tt>extra.parameter()</tt> function as follows.

``` r
fitPL_PE_start <- PLreg(percentfat ~ days + sex + year, data = bodyfat_Aeolus,
              family = "PE", zeta = 2)
extra.parameter(fitPL_PE_start, lower = 1, upper = 2.5)
```

Then, fit the model with the chosen value for
![\\zeta](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Czeta "\zeta").

``` r
fitPL_PE <- PLreg(percentfat ~ days + sex + year, data = bodyfat_Aeolus,
              family = "PE", zeta = 1.7)
summary(fitPL_PE)
#> 
#> Call:
#> PLreg(formula = percentfat ~ days + sex + year, data = bodyfat_Aeolus, 
#>     family = "PE", zeta = 1.7)
#> 
#> Standardized residuals:
#>     Min      1Q  Median      3Q     Max 
#> -3.0006 -0.6000  0.0224  0.6353  2.5352 
#> 
#> Coefficients (median model with logit link):
#>               Estimate Std. Error z value Pr(>|z|)    
#> (Intercept) -1.1615018  0.0647074 -17.950   <2e-16 ***
#> days        -0.0092638  0.0005363 -17.273   <2e-16 ***
#> sexM        -0.0499217  0.0570871  -0.874    0.382    
#> year2016     0.5176694  0.0607028   8.528   <2e-16 ***
#> 
#> Sigma coefficients (dispersion model with log link):
#>         Estimate Std. Error z value Pr(>|z|)    
#> (sigma) -1.94446    0.07869  -24.71   <2e-16 ***
#> 
#> Lambda coefficient:
#>           Estimate Std. Error
#> (lambda) 0.0004124      0.067
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1 
#> 
#> Family: PL - PE ( 1.7 ) (Power logit power exponential)
#> Estimation method: pML (penalized maximum likelihood)
#> Log-likelihood:   320 on 6 Df
#> Pseudo R-squared: 0.6756
#> Upsilon statistic: 0.04852
#> AIC:  -628
#> Number of iterations in BFGS optimization: 20
```

The goodness of fit is assessed using diagnostic graphs through the plot
method.

``` r
plot(fitPL_PE, which = 1:4)
```

Further details and examples on the R package
![\\textbf{PLreg}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Ctextbf%7BPLreg%7D "\textbf{PLreg}")
can be found using the help on R by typing:

    help("PLreg")

## Reference

Queiroz, F. F. and Ferrari, S. L. P. (2022). Power logit regression for
modeling bounded data.
![\\textit{arXiv}](https://latex.codecogs.com/png.image?%5Cdpi%7B110%7D&space;%5Cbg_white&space;%5Ctextit%7BarXiv%7D "\textit{arXiv}"):2202.01697.


NEWS.md
# PLreg 0.4.1

## Minor changes

* Small changes in the documentation of some functions.

## Miner changes

* Update the CI.lambda function.

tests/testthat.R
library(testthat)
library(PLreg)

test_check("PLreg")


tests/testthat/test-PLreg.R
#### IID case

mu <- 0.3; sigma <- 1; lambda <- 2
set.seed(1)
PLsample <- rPL(1000, mu, sigma, lambda, family = "SN",  zeta = 2.5)
hist(PLsample, prob = TRUE, breaks = 15, main = "", las = 1)
curve(dPL(x, mu, sigma, lambda, family = "SN",  zeta = 2.5),
      from = 0.01, to = 0.8, add = TRUE, col = "red")
rug(PLsample)

x <- seq(0.01, 0.9,0.01)
y <- dPL(x, mu, sigma, lambda, family = "Hyp",  zeta = 2)
plot(x, y, type = "l", lwd = 2, las = 1)

x1 <- seq(0.01, 0.4, 0.01)
y1 <- dPL(x1, mu, sigma, lambda, family = "Hyp",  zeta = 2)
polygon(c(x1, 0.4, 0), c(y1, 0, 0), col = "lightblue")

text(mu-0.025, 1, paste("P(Y<0.4) = ",
                       round(pPL(0.4, mu, sigma, lambda, family = "Hyp",  zeta = 2),2)),
     font = 1, cex = 0.8)

plot(x, pPL(x, mu, sigma, lambda, family = "PE",  zeta = 1.3),
     type = "l", las = 1, lwd = 2, ylab = expression(P(Y<y)), xlab = "y")
p <- pPL(0.5, mu, sigma, lambda, family = "PE",  zeta = 1.3)
q <- qPL(p, mu, sigma, lambda, family = "PE",  zeta = 1.3)
points(q, p, pch = 16, col = 2, cex = 1.5)
text(0.55, 0.83, paste("(", 0.5, ",", round(p, 2), ")"), font = 2,
     cex = 0.8, col = "red")


#### IID case
data("bodyfat_Aeolus")

fitPL <- PLreg(percentfat ~ 1, data = bodyfat_Aeolus,
               family = "SN", zeta = 1.6)
summary(fitPL)

extra.parameter(fitPL, lower = 1, upper = 4)


hist(bodyfat_Aeolus$percentfat, prob = TRUE)
curve(dPL(x, mu = fitPL$fitted.values[1],
          sigma = exp(fitPL$coefficients$dispersion),
          lambda = fitPL$coefficients$skewness,
          zeta = fitPL$zeta, family = fitPL$family),
      from = 0.001, to = 0.4, add = TRUE, col = "red")



plot(ecdf(bodyfat_Aeolus$percentfat),verticals = TRUE,  lwd = 2, col = "gray",
     main = "", ylab="Cumulative density function", xlab = "y")
stripchart(bodyfat_Aeolus$percentfat, add = TRUE, at = 0, col = "black", pch=3)
curve(pPL(x, mu = fitPL$fitted.values[1],
          sigma = exp(fitPL$coefficients$dispersion),
          lambda = fitPL$coefficients$skewness,
          zeta = fitPL$zeta, family = fitPL$family),
      from = 0.001, to = 0.4, add = TRUE, col = "red")




#### Errors


expect_error(PLreg(percentfat ~ 1, data = bodyfat_Aeolus,
                   family = "normal",
                   zeta = 4))
expect_error(PLreg(percentfat ~ 1, data = bodyfat_Aeolus, family = "TF"))
expect_error(PLreg(percentfat ~ 1, family = "TF"))
expect_error(PLreg(percentfat ~ 1, data = bodyfat_Aeolus,
               family = "TF", zeta = 0))
expect_error(summary(PLreg(percentfat ~ 1, data = bodyfat_Aeolus,
               family = "TF", zeta = 2)))
fitPL <- PLreg(percentfat ~ 1, data = bodyfat_Aeolus,
               family = "NO")
expect_error(extra.parameter(fitPL, lower = 1, upper = 4))

#### Different link functions
data("PeruVotes")
fitPL <- PLreg(votes ~ HDI | HDI, data = PeruVotes,
               family = "TF", zeta = 5,
               link = "probit", link.sigma = "sqrt")
summary(fitPL)

#### Different residuals
data("PeruVotes")
fitPL <- PLreg(votes ~ HDI | HDI, data = PeruVotes,
               family = "TF", zeta = 5)

res_quantile <- residuals(fitPL, type = "quantile")
res_standardized <- residuals(fitPL, type = "standardized")

plot(res_standardized, pch = "+", ylim = c(-6, 6))
abline(h = -3, lty = 2)
abline(h = 3, lty = 2)

qqnorm(res_quantile)
qqline(res_quantile)

#### Body fat data
data("bodyfat_Aeolus")


#Initial model with zeta = 2
fit1 <- PLreg(percentfat ~ days + sex + year, data = bodyfat_Aeolus,
                     family = "PE", zeta = 2)
summary(fit1)
# Choosing the best value for zeta
extra.parameter(fit1, lower = 1, upper = 4, grid = 15)

# Using zeta = 1.7
fit2 <- PLreg(percentfat ~ days + sex + year, data = bodyfat_Aeolus,
              family = "PE", zeta = 1.7)
summary(fit2)

# Fixing lambda = 1
fit3 <- PLreg(percentfat ~ days + sex + year, data = bodyfat_Aeolus,
              family = "PE", zeta = 1.7, control = PLreg.control(lambda = 1))
summary(fit3)

# Comparing the AIC and Upsilon values between fit2 and fit3
AIC(fit2) < AIC(fit3)
fit2$Upsilon.zeta < fit3$Upsilon.zeta


#### Peru votes
data("PeruVotes")

fitPL <- PLreg(votes ~ HDI | HDI, data = PeruVotes,
               family = "TF", zeta = 5, control = PLreg.control(lambda = 1))
summary(fitPL)



plot(fitPL, type = "standardized")

#### Firm
data("Firm")

fitPL <- PLreg(firmcost ~ sizelog + indcost | sizelog + indcost, data = Firm,
               family = "SLASH", zeta = 2.13)
summary(fitPL)
extra.parameter(fitPL, lower = 1.2, upper = 4, grid = 10)
plot(fitPL, type = "standardized")
envelope(fitPL, type = "standardized")

fitPL_wo72 <- PLreg(firmcost ~ sizelog + indcost | sizelog + indcost, data = Firm[-72,],
               family = "SLASH", zeta = 2.13)
fitPL_wo15 <- PLreg(firmcost ~ sizelog + indcost | sizelog + indcost, data = Firm[-15,],
                    family = "SLASH", zeta = 2.13)
fitPL_wo16 <- PLreg(firmcost ~ sizelog + indcost | sizelog + indcost, data = Firm[-16,],
                    family = "SLASH", zeta = 2.13)

coef.mu      <- coef(fitPL)[1:3]
coef.mu_wo72 <- coef(fitPL_wo72)[1:3]
coef.mu_wo15 <- coef(fitPL_wo15)[1:3]
coef.mu_wo16 <- coef(fitPL_wo16)[1:3]

plot(Firm$indcost, Firm$firmcost, pch = "+", xlab = "indcost", ylab = "firmcost")
#identify(Firm$indcost, Firm$firmcost)
covariate <- matrix(c(rep.int(1, 1000), rep(median(Firm$sizelog), 1000),
                     seq(0, 1.22, length.out = 1000)),
                   ncol = 3)
lines(covariate[,3], as.vector(fitPL$link$median$linkinv(covariate%*%coef.mu)),
      type = "l")
lines(covariate[,3], as.vector(fitPL$link$median$linkinv(covariate%*%coef.mu_wo72)),
      type = "l", lty = 2, col = "blue")
lines(covariate[,3], as.vector(fitPL$link$median$linkinv(covariate%*%coef.mu_wo15)),
      type = "l", lty = 3, col = "red")
lines(covariate[,3], as.vector(fitPL$link$median$linkinv(covariate%*%coef.mu_wo16)),
      type = "l", lty = 4, col = "green")
parameters <- c("pML",
             "pML w/o 72",
             "pML w/o 15",
             "pML w/o 16")
legend(x = 0.5,
       y = 0.8,
       legend = parameters,
       col = c("black", "blue", "red", "green"),
       lty = c(1, 2, 3, 4),
       cex = 0.6)



#### influence diagnostics
data("Firm")

fitPL <- PLreg(firmcost ~ sizelog + indcost | sizelog + indcost, data = Firm,
               family = "SLASH", zeta = 2.13)

influence_measures <- influence(fitPL, graph = FALSE)
plot(influence_measures$case.weights, type = "h", ylim = c(0,1))
plot(influence_measures$totalLI, type = "h", ylim = c(0,6))
plot(Firm$sizelog, influence_measures$GL, pch = "+")


### Plot function

data("Firm")

fitPL <- PLreg(firmcost ~ sizelog + indcost | sizelog + indcost, data = Firm,
               family = "NO", zeta = 2.13)
par(mfrow = c(3,3))
plot(fitPL, type = "standardized")
par(mfrow = c(1, 1))


tests/testthat/test-PLdist.R



test_that("GJS dist with mu equal 0.5 is symetric", {
  expect_equal(pPL(0.5, mu = 0.5, sigma = 2, lambda = 1, family = "NO"), 0.5)
  expect_equal(pPL(0.5, mu = 0.5, sigma = 2, lambda = 1, family = "SLASH"), 0.5)
  expect_equal(pPL(0.5, mu = 0.5, sigma = 2, lambda = 1, family = "LO"), 0.5)
})


R/PeruVotes.R
#' Peru Blank Votes
#'
#' A dataset on the blank votes in the 2006 Peruvian general election.
#'
#' @usage data(PeruVotes)
#' @format A data frame with 194 rows and 2 variables:
#' \describe{
#'   \item{votes}{proportion of blank votes in the 2006 Peruvian general election.}
#'   \item{HDI}{Human Development Index.}
#' }
#' @details The dataset was collected by Bayes et al. (2012) and the response variable
#'     is \code{votes}, proportion of blank votes in the 2006 Peruvian general election.
#'     Bayes et al. (2012) analyzed the influence of the Human Development Index (HDI) on
#'     the proportion of blank votes using a beta rectangular regression model. Lemonte and
#'     Bazan (2015) also analyzed the data using GJS regression models.
#' @keywords datasets
#' @source \url{https://www.undp.org/es/peru}
#' @references Bayes, C., Bazan, J. L. and Garcia, C. (2012). A new robust regression model for proportions.
#'     \emph{Bayesian Analysis}. 7:771-796 \cr \cr
#'     Lemonte, A. J. and Bazan, J. L. (2015). New class of Johnson SB distributions
#'     and its associated regression model for rates and proportions. \emph{Biometrical Journal}. 58:727-746.
#' @examples data("PeruVotes")
#' fitPL <- PLreg(votes ~ HDI | HDI,
#'                data = PeruVotes,
#'                family = "TF",
#'                zeta = 5,
#'                control = PLreg.control(lambda = 1))
#' summary(fitPL)
#' plot(fitPL, type = "standardized")
"PeruVotes"


data-raw/bodyfat_Aeolus.R
## code to prepare `bodyfat_Aeolus` dataset goes here

data <- read.csv("data-raw/ChengJAE2018_data4dryad.csv", header = TRUE, sep = ";", dec = ".")
data <- data[, -c(1,4,6,7,8,9,10)]

year  <- c(rep("2016", 239), rep("2009", 350))
data  <- cbind(data, year)

data <- na.omit(data)
rownames(data) <- NULL

aux <- data[data$site == "Aeolus", -c(1)]
rownames(aux) <- NULL

days <- NULL
for(i in 1:length(aux[,1])){
  if( aux$date[i] == "2015-Nov-17"){
    days[i] <- 127
  }
  if( aux$date[i] == "2015-Nov-13"){
    days[i] <- 130
  }
  if( aux$date[i] == "2016-Mar-09"){
    days[i] <- 15
  }
  if( aux$date[i] == "2008-Nov-18"){
    days[i] <- 126
  }
  if( aux$date[i] == "2009-Jan-31"){
    days[i] <- 52
  }
  if( aux$date[i] == "2009-Mar-27"){
    days[i] <- 0
  }
  if( aux$date[i] == "2015-Nov-05"){
    days[i] <- 139
  }
  if( aux$date[i] == "2016-Mar-10"){
    days[i] <- 13
  }
  if( aux$date[i] == "2008-Nov-18"){
    days[i] <- 126
  }
  if( aux$date[i] == "2009-Jan-12"){
    days[i] <- 71
  }
  if( aux$date[i] == "2009-Mar-19"){
    days[i] <- 4
  }
  if( aux$date[i] == "2015-Nov-06"){
    days[i] <- 138
  }
  if( aux$date[i] == "2015-Nov-05"){
    days[i] <- 139
  }
  if( aux$date[i] == "2016-Mar-04"){
    days[i] <- 20
  }
  if( aux$date[i] == "2016-Feb-17"){
    days[i] <- 35
  }
  if( aux$date[i] == "2008-Feb-12"){
    days[i] <- 40
  }
  if( aux$date[i] == "2008-Mar-26"){
    days[i] <- 0
  }
  if( aux$date[i] == "2015-Nov-18"){
    days[i] <- 126
  }
  if( aux$date[i] == "2016-Mar-21"){
    days[i] <- 2
  }
  if( aux$date[i] == "2008-Nov-25"){
    days[i] <- 119
  }
  if( aux$date[i] == "2009-Jan-13"){
    days[i] <- 70
  }
  if( aux$date[i] == "2009-Mar-15"){
    days[i] <- 8
  }
  if( aux$date[i] == "2009-Mar-16"){
    days[i] <- 7
  }
  if( aux$date[i] == "2008-Dec-17"){
    days[i] <- 97
  }
}

days <- 175 - days
bodyfat_Aeolus <- cbind(aux[,-1], days)
colnames(bodyfat_Aeolus) <- c("sex", "percentfat", "year", "days")

bodyfat_Aeolus$sex <- factor(bodyfat_Aeolus$sex)
bodyfat_Aeolus$year <- factor(bodyfat_Aeolus$year)

usethis::use_data(bodyfat_Aeolus, overwrite = TRUE)


R/PLdist.R
#' @keywords internal
pslash <- function(q, nu){
  aux <-  function(x){
    s_aux <- x^2/2
    beta_aux <- nu + (1/2)
    gama_aux <- zipfR::Igamma(beta_aux, s_aux)
    r <- (nu/sqrt(2*pi))*(1/(s_aux^beta_aux))*zipfR::Igamma(beta_aux, s_aux)
    return(r)
  }

  acumu_aux <- function(q) stats::integrate(aux, lower=-Inf, upper=q)$value
  v.acumu_aux <- Vectorize(acumu_aux)
  cdf <- v.acumu_aux(q)
  cdf
}

#' @keywords internal
qslash <- function(p, nu){

  qtf <- function(input){
    p <- input

    if (!is.na(p)){
      obj <- function(q){
        pslash(q, nu) - p
      }

      nleqslv::nleqslv(stats::qnorm(p) / (0.5^(1/nu)), obj)$x
    }else{
      numeric(0)
    }
  }

  q0 <- rep(0, length(p[p == 0.5]) )
  q <- vector()
  if (length(p[p != 0.5]) < 1){
    q <- numeric(0)
  } else{
    q[p != 0.5 & p != 0 & p != 1] <-
      as.numeric(apply(matrix(p[p != 0.5 & p != 0 & p != 1], ncol = 1),
                       1, qtf))
  }

  q[p == 0] <- -Inf
  q[p == 1] <- Inf
  qtf <- c(q0, q)
  index <- c(which(p == 0.5), which(p != 0.5))

  qtf[sort(index, index.return = T)$ix]
}


#' Power Logit Distributions
#'
#' Density, distribution function, quantile function and random generation
#' for power logit distributions.

#' @name PL
#' @param x,q vector of quantiles.
#' @param p vector of probabilities.
#' @param n number of observations. If \code{length(n) > 1}, the length is taken to be the number required.
#' @param mu vector of medians.
#' @param sigma vector of dispersion parameters.
#' @param lambda vector of skewness parameters.
#' @param zeta vector of extra parameters.
#' @param family string that specifies the family used to define the power logit distribution. The family is
#'     \code{NO}, \code{TF}, \code{LO}, \code{PE}, \code{SHN}, \code{Hyp} and \code{SLASH} for normal,
#'     Student-t, type II logistic, power exponential, sinh-normal, hyperbolic and slash
#'     distribution, respectively.
#' @param log,log.p logical; if TRUE, probabilities p are given as log(p). Default is FALSE.
#' @param lower.tail logical; if TRUE (default), probabilities are \eqn{P(X \le x)} otherwise, \eqn{P(X > x)}.
#'
#' @details If \code{zeta} is not specified, it assumes the default value 2. \cr
#'     The power logit distribution has density \deqn{f(y; \mu, \sigma, \lambda) = \lambda r(z^2)/(\sigma y (1-y^\lambda)),}
#'     for \eqn{y \in (0,1)}, in which \eqn{z = [logit(y^\lambda) - logit(\mu^\lambda)]/\sigma}, \eqn{r(\cdot)} is the density generator
#'     and \eqn{\mu \in (0,1)}, \eqn{\sigma>0} and \eqn{\lambda>0} are the median, dispersion and skewness of the distribution. \cr
#'     It is possible to consider \eqn{\lambda=0} to obtain the limiting case, the log-log distribution. This distribution has
#'     density \deqn{f(y; \mu, \sigma, \lambda) = r(z^2)/(\sigma y (-log(y))),} for \eqn{y \in (0,1)},
#'     in which \eqn{z = [-log(-log(y)) - (-log(-log(y)))]/\sigma}. \cr
#'     The \code{family} argument defines the density generator \eqn{r(\cdot)}, which may depend on an extra parameter (\code{zeta}).
#' @return \code{dPL} gives the density, \code{pPL} gives the distribution function, \code{qPL} gives the quantile function,
#'     and \code{rPL} generates random variables.
#' @references Queiroz, F. F. and Ferrari, S. L. P. (2022). Power logit regression 
#'       for modeling bounded data. \emph{arXiv preprint arXiv}:2202.01697.
#' @examples
#' dPL(0.2, mu = 0.3, sigma = 1, lambda=1, zeta = 2, family = "PE")
#' mu = 0.3; sigma = 1; lambda = 2
#' set.seed(1)
#' PLsample = rPL(1000, mu, sigma, lambda, family = "SN",  zeta = 2.5)
#' hist(PLsample, prob = TRUE, breaks = 15, main = "", las = 1)
#' curve(dPL(x, mu, sigma, lambda, family = "SN",  zeta = 2.5),
#'       from = 0.01, to = 0.8, add = TRUE, col = "red")
#' rug(PLsample)
#'
#' x = seq(0.01, 0.9,0.01)
#' y = dPL(x, mu, sigma, lambda, family = "Hyp",  zeta = 2)
#' plot(x, y, type = "l", lwd = 2, las = 1)
#'
#' x1 = seq(0.01, 0.4, 0.01)
#' y1 = dPL(x1, mu, sigma, lambda, family = "Hyp",  zeta = 2)
#' polygon(c(x1, 0.4, 0), c(y1, 0, 0), col = "lightblue")
#' text(mu-0.025, 1, paste("P(Y<0.4) = ",
#'                         round(pPL(0.4, mu, sigma, lambda,
#'                         family = "Hyp",
#'                         zeta = 2),2)),
#'      font = 1, cex = 0.8)
#'
#' plot(x, pPL(x, mu, sigma, lambda, family = "PE",  zeta = 1.3),
#'      type = "l", las = 1, lwd = 2,
#'      ylab = expression(P(Y<y)),
#'      xlab = "y")
#' p = pPL(0.5, mu, sigma, lambda, family = "PE",  zeta = 1.3)
#' q = qPL(p, mu, sigma, lambda, family = "PE",  zeta = 1.3)
#' points(q, p, pch = 16, col = 2, cex = 1.5)
#' text(0.55, 0.83, paste("(", 0.5, ",", round(p, 2), ")"), font = 2,
#'      cex = 0.8, col = "red")
#' @importFrom stats dnorm qnorm rnorm pnorm dlogis plogis qlogis runif dt pt qt rbeta qbeta
#' @export
dPL <- function(x, mu, sigma, lambda, zeta = 2, family, log = FALSE){

  if (any(mu <= 0) | any(mu >= 1))
    stop(paste("mu must be beetwen 0 and 1 ", "\n", ""))
  if (any(sigma <= 0))
    stop(paste("sigma must be positive", "\n", ""))
  if (any(lambda < 0))
    stop(paste("lambda must be greater than or equal to 0", "\n", ""))
  if (any(zeta <= 0))
    stop(paste("zeta must be positive ", "\n", ""))
  if (any(x < 0) | any(x > 1))
    stop(paste("x must be beetwen (0, 1)", "\n", ""))

  if(lambda == 0){
    z <- (-1/sigma)*(log(log(x)/log(mu)))
  }else{
    z <- (1/sigma)*(VGAM::logitlink(x^lambda) - VGAM::logitlink(mu^lambda))
  }

  if(family == "NO"){
    r <- dnorm(z)
  }
  if(family == "TF"){
    r <- dt(z, zeta)
  }
  if(family == "LO"){
    r <- dlogis(z)
  }
  if(family == "SN"){
    r <- (2*cosh(sqrt(z^2))*exp(-2*sinh(sqrt(z^2))*sinh(sqrt(z^2))/zeta^2)/(sqrt(2*pi)*zeta))
  }
  if(family == "PE"){
    r <- gamlss.dist::dPE(z, mu = 0, sigma = 1, nu = zeta)
  }
  if(family == "Hyp"){
    r <- GeneralizedHyperbolic::dhyperb(z, mu = 0, delta = 1, alpha = zeta, beta = 0)
  }
  if(family == "SLASH"){
    s_aux <- z^2/2
    beta_aux <- zeta + (1/2)
    if(any(s_aux == 0)){
      s_aux[s_aux == 0] <- 0.0001
    }
    r <- (zeta/sqrt(2*pi))*(1/(s_aux^beta_aux))*zipfR::Igamma(beta_aux, s_aux)
  }

  if(lambda == 0){
    log.lik <- ifelse(x <= 0 | x >= 1, -Inf, - log(abs(sigma*x*log(x))) + log(r))
  }else{
    log.lik <- ifelse(x <= 0 | x >= 1, -Inf, log(lambda) - (log(sigma) + log(1 - x^lambda) + log(x) - log(r)))
  }

  if (log == FALSE)
    fy <- exp(log.lik)
  else fy <- log.lik
  fy
}

#' @rdname PL
#' @export
pPL <- function(q, mu, sigma, lambda, zeta = 2, family, lower.tail = TRUE, log.p = FALSE){


  if (any(mu <= 0) | any(mu >= 1))
    stop(paste("mu must be beetwen 0 and 1 ", "\n", ""))
  if (any(sigma <= 0))
    stop(paste("sigma must be positive", "\n", ""))
  if (any(lambda < 0))
    stop(paste("lambda must be greater than or equal to 0", "\n", ""))
  if (any(zeta <= 0))
    stop(paste("zeta must be positive ", "\n", ""))

  if(lambda == 0){
    z <- (-1/sigma)*(log(log(q)/log(mu)))
  }else{
    z <- (1/sigma)*(VGAM::logitlink(q^lambda) - VGAM::logitlink(mu^lambda))
  }

  if(family == "NO"){
    cdf <- pnorm(z, lower.tail = TRUE, log.p = FALSE)
  }
  if(family == "TF"){
    cdf <- pt(z, zeta, lower.tail = TRUE, log.p = FALSE)
  }
  if(family == "LO"){
    cdf <- plogis(z, lower.tail = TRUE, log.p = FALSE)
  }
  if(family == "SN"){
    cdf <-  pnorm((2/zeta)*sinh(z), lower.tail = TRUE, log.p = FALSE)
  }
  if(family == "PE"){
    cdf <- gamlss.dist::pPE(z, mu = 0, sigma = 1, nu = zeta, lower.tail = TRUE, log.p = FALSE)
  }
  if(family == "Hyp"){
    cdf <- GeneralizedHyperbolic::phyperb(z, mu = 0, delta = 1, alpha = zeta, beta = 0)
  }
  if(family == "SLASH"){
    cdf <- pslash(z, nu = zeta)
  }

  if (lower.tail == TRUE)
    cdf <- cdf
  else cdf <- 1 - cdf
  if (log.p == FALSE)
    cdf <- cdf
  else cdf <- log(cdf)
  cdf
}

#' @rdname PL
#' @export
qPL <- function(p, mu, sigma, lambda, zeta = 2, family, lower.tail = TRUE, log.p = FALSE){

  if (any(mu <= 0) | any(mu >= 1))
    stop(paste("mu must be beetwen 0 and 1 ", "\n", ""))
  if (any(sigma <= 0))
    stop(paste("sigma must be positive", "\n", ""))
  if (any(lambda < 0))
    stop(paste("lambda must be greater than or equal to 0", "\n", ""))
  if (any(zeta <= 0))
    stop(paste("zeta must be positive ", "\n", ""))

  if (log.p == TRUE)
    p <- exp(p)
  else p <- p
  if (lower.tail == TRUE)
    p <- p
  else p <- 1 - p
  if (any(p < 0) | any(p > 1))
    stop(paste("p must be between 0 and 1", "\n", ""))

  if(family == "NO"){
    z_p <- qnorm(p, lower.tail = TRUE)
  }
  if(family == "TF"){
    z_p <- qt(p, zeta, lower.tail = TRUE)
  }
  if(family == "LO"){
    z_p <- qlogis(p, lower.tail = TRUE)
  }
  if(family == "SN"){
    z_p <- asinh((zeta/2)*qnorm(p, lower.tail = TRUE))
  }
  if(family == "PE"){
    z_p <- gamlss.dist::qPE(p, mu = 0, sigma = 1, nu = zeta, lower.tail = TRUE)
  }
  if(family == "Hyp"){
    z_p <- GeneralizedHyperbolic::qhyperb(p, mu = 0, delta = 1, alpha = zeta, beta = 0, lower.tail = TRUE)
  }
  if(family == "SLASH"){
    z_p <- qslash(p, nu = zeta)
  }

  if(lambda == 0){
    suppressWarnings(q <- exp(log(mu)*exp(-sigma*z_p)))
  }else{
    suppressWarnings(q <- ((mu^lambda)*exp(sigma*z_p)/(1 - (mu^lambda)*(1 - exp(sigma*z_p))))^(1/lambda))
  }
  q
}

#' @rdname PL
#' @export
rPL <- function(n, mu, sigma, lambda, zeta = 2, family){

  if (any(mu <= 0) | any(mu >= 1))
    stop(paste("mu must be beetwen 0 and 1 ", "\n", ""))
  if (any(sigma <= 0))
    stop(paste("sigma must be positive", "\n", ""))
  if (any(lambda < 0))
    stop(paste("lambda must be greater than or equal to 0", "\n", ""))
  if (any(zeta <= 0))
    stop(paste("zeta must be positive ", "\n", ""))
  if (any(n <= 0))
    stop(paste("n must be a positive integer", "\n", ""))

  n <- ceiling(n)

  p <- runif(n)
  if(family == "SLASH"){
    z_p <- (qbeta(p, zeta, 1, lower.tail = TRUE)^(-1/2))*qnorm(p, lower.tail = TRUE)
    if(lambda == 0){
      suppressWarnings(r <- exp(log(mu)*exp(-sigma*z_p)))
    }else{
      suppressWarnings(r <- ((mu^lambda)*exp(sigma*z_p)/(1 - (mu^lambda)*(1 - exp(sigma*z_p))))^(1/lambda))
    }
  }else{
    r <- qPL(p, mu = mu, sigma = sigma, lambda = lambda, zeta = zeta, family = family)
  }
  r
}


